lib/reporting/monthly_idv_report.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+require 'csv'
+require 'reporting/identity_verification_report'
+require 'reporting/unknown_progress_bar'
+
+module Reporting
+  class MonthlyIdvReport
+    attr_reader :end_date
+
+    def initialize(end_date:, verbose: false, progress: false, parallel: false)
+      @end_date = end_date.in_time_zone('UTC')
+      @verbose = verbose
+      @progress = progress
+      @parallel = parallel
+    end
+
+    # Make all of these actually configurable...
+    def verbose?
+      @verbose
+    end
+
+    def progress?
+      @progress
+    end
+
+    def parallel?
+      @parallel
+    end
+
+    def monthly_idv_report_emailable_report
+      EmailableReport.new(
+        title: 'Proofing Rate Metrics',
+        subtitle: 'Condensed 🆕',
+        float_as_percent: true,
+        precision: 2,
+        table: as_csv,
+        filename: 'condensed_hop_results', # FIXME, ask May
+      )
+    end
+
+    # Should I cargo cult in the ThrottlingException from above? Can we extract that?
+    def as_csv
+      csv = []
+
+      csv << ['Metric', *reports.map { |t| t.time_range.begin.strftime('%b %Y') }]
+      csv << ['IDV started', *reports.map(&:idv_started)]
+
+      csv << ['# of successfully verified users', *reports.map(&:successfully_verified_users)]
+      csv << ['% IDV started to successfully verified', *reports.map(&:blanket_proofing_rates)]
+
+      csv << ['# of workflow completed', *reports.map(&:idv_final_resolution)]
+      csv << ['% rate of workflow completed', *reports.map(&:idv_final_resolution_rate)]
+
+      csv << ['# of users verified (total)', *reports.map(&:verified_user_count)]
+    rescue Aws::CloudWatchLogs::Errors::ThrottlingException => err
+      [
+        ['Error', 'Message'],
+        [err.class.name, err.message],
+      ]
+    end
+
+    def reports
+      @reports ||= begin
+        parallel? ? parallel_reports : non_parallel_reports
+      end
+    end
+
+    def non_parallel_reports
+      Reporting::UnknownProgressBar.wrap(show_bar: progress?) do
+        monthly_subreports.each(&:data)
+      end
+    end
+
+    def parallel_reports
+      threads = monthly_subreports.map do |report|
+        Thread.new do
+          report.tap(&:data)
+        end.tap do |thread|
+          thread.report_on_exception = false
+        end
+      end
+
+      Reporting::UnknownProgressBar.wrap(show_bar: progress?) do
+        threads.map(&:value)
+      end
+    end
+
+    def monthly_subreports
+      ranges = [
+        (end_date - 3.months).all_month,
+        (end_date - 2.months).all_month,
+        (end_date - 1.month).all_month,
+      ]
```

Comment:
🤔❓ 

I have changed this block repeatedly in terms of which time window it looks at.

When we go to send the report on September 1, we want to include the full months of June, July, and August.

If we include this in the daily email once this is merged, the run on, e.g., August 20 would include May, June, and July, and because it's looking at CloudWatch events for a given time window, it would always be the same every day. (Except for the last row that's a DB query, which may vary a little as profiles are deactivated.)

It might be more interesting if, for the non-monthly ones, we included the present month. But every bone in my body thinks changing how the report is calculated based on the day is how you get embarrassing bugs where it looks fine until the report goes out to leadership.

It's also really wasteful to recalculate this every day? It took about 20 minutes (60 with parallel: false) on my laptop; it might be a little more expeditious, but not tremendously so, running from within AWS. I almost want to add some caching layer, but talk about scope creep...

Reply:
We run the reports at midnight-ish with `Time.zone.yesterday.end_of_day`

So on September 1, we'll get:
```ruby
> end_date = Date.new(2024, 9, 1).yesterday.end_of_day
=> Sat, 31 Aug 2024 23:59:59.999999999 UTC +00:00
```

so these range will be May, June, July which is not what we want.

```ruby
=> [Wed, 01 May 2024 00:00:00.000000000 UTC +00:00..Fri, 31 May 2024 23:59:59.999999999 UTC +00:00,
 Sat, 01 Jun 2024 00:00:00.000000000 UTC +00:00..Sun, 30 Jun 2024 23:59:59.999999999 UTC +00:00,
 Mon, 01 Jul 2024 00:00:00.000000000 UTC +00:00..Wed, 31 Jul 2024 23:59:59.999999999 UTC +00:00]
```

```suggestion
      ranges = [
        (end_date - 2.months).all_month,
        (end_date - 1.month).all_month,
        end_date.all_month,
      ]
```